compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -578,6 +578,172 @@ static int64_t getMaxVectorTileSize(int64_t numElem, int64_t tileSize,
   return 1;
 }
 
+static bool isInnerMostDimThatMapIsFunctionOf(AffineMap map, int dim) {
+  if (!map.isFunctionOfDim(dim)) {
+    return false;
+  }
+  for (int d = dim + 1; d < map.getNumDims(); ++d) {
+    if (map.isFunctionOfDim(d)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+static int sum(ArrayRef<int> a) {
+  int s = 0;
+  for (int i : a) {
+    s += i;
+  }
+  return s;
+}
+
+// Clamps in-place `vecTileSizes`, ensuring that the resulting vector tile sizes
+// for each opearand of `op` satisfy two requirements:
+// 1. No resulting operand tile size exceeds `eachOperandMaxTileBits`.
+// 2. The sum of all resulting operand tile sizes does not exceed
+// `allOperandsMaxTileBits`.
+static void limitVectorTileSizes(linalg::LinalgOp op,
+                                 SmallVectorImpl<int64_t> &vecTileSizes,
+                                 int64_t eachOperandMaxTileBits,
+                                 int64_t allOperandsMaxTileBits) {
+  int numLoops = op.getNumLoops();
+  assert(numLoops == vecTileSizes.size());
+  auto indexingMaps = op.getIndexingMapsArray();
+  auto operandTypes = op->getOperandTypes();
+  int numOperands = operandTypes.size();
+
+  // For each operand, we track how big the tile is going to be based on the
+  // dimensions that we have already accounted for. Here we initialize this to
+  // just elemBitWidth, having not yet accounted for any array dimension.
+  SmallVector<int> tileBits(numOperands);
+  for (int i : llvm::seq<int>(0, numOperands)) {
+    tileBits[i] = operandTypes[i].cast<ShapedType>().getElementTypeBitWidth();
+    if (tileBits[i] > eachOperandMaxTileBits) {
+      // eachOperandMaxTileBits is too small for this requirement to be
+      // satisfied. Early-returning. This might conceivably happen due to the
+      // tail-recursion in this function, where eachOperandMaxTileBits is
+      // halved.
+      return;
+    }
+  }
+
+  // For each loop, we must make sure that if it's the inner-most dimension that
+  // concerns an operand with a less-than-8-bit element type, we preserve the
+  // requirement that the inner-most tile dimension is >= 8 bit. For example, if
+  // the element type is `i2`, the minimum inner tile size is 8/2 = 4.
+  SmallVector<int64_t> minVecTileSizes(numLoops);
+  for (int loopNum : llvm::seq<int>(0, numLoops)) {
+    minVecTileSizes[loopNum] = 1;
+    for (int i : llvm::seq<int>(0, numOperands)) {
+      if (isInnerMostDimThatMapIsFunctionOf(indexingMaps[i], loopNum)) {
+        int64_t elemBitWidth =
+            operandTypes[i].cast<ShapedType>().getElementTypeBitWidth();
+        if (elemBitWidth < 8) {
+          minVecTileSizes[loopNum] =
+              std::max(minVecTileSizes[loopNum], 8 / elemBitWidth);
```

Comment:
Good catch! Fixed by taking the greatest common divisor of the bit-width and 8.

For example, if the element type if i6, since gcd(8, 6) = 2, we are in the same situation as with `i2` and the minimum inner tile size is again 8/2 = 4.  Indeed `4xi6` is 24 bits is a multiple of 8 bits.